### PR TITLE
Client Certificate Support for Schannel

### PR DESCRIPTION
**Pull request by anrossi**
Initial client certificate support for Schannel in user mode on Windows.  Includes a test that only builds for user mode Schannel.
Further tests and more platform support will come in subsequent PRs.

### CODE REVIEW

File: src/perf/bin/appmain.cpp [hunk 1]
```diff
@@ -393,7 +393,8 @@ main(
         CxPlatPlatGetSelfSignedCert(
             TestingKernelMode ?
                 CXPLAT_SELF_SIGN_CERT_MACHINE :
-                CXPLAT_SELF_SIGN_CERT_USER);
+                CXPLAT_SELF_SIGN_CERT_USER,
+        FALSE);
```
Comment (nibanks) [inline src/perf/bin/appmain.cpp:397]: nit: indent
```suggestion
            FALSE);
```

File: src/platform/selfsign_openssl.c [hunk 2]
```diff
@@ -297,10 +297,12 @@ typedef struct CXPLAT_CREDENTIAL_CONFIG_INTERNAL {
 _IRQL_requires_max_(PASSIVE_LEVEL)
 const QUIC_CREDENTIAL_CONFIG*
 CxPlatPlatGetSelfSignedCert(
-    _In_ CXPLAT_SELF_SIGN_CERT_TYPE Type
+    _In_ CXPLAT_SELF_SIGN_CERT_TYPE Type,
+    _In_ BOOLEAN ClientCertificate
     )
 {
     UNREFERENCED_PARAMETER(Type);
+    UNREFERENCED_PARAMETER(ClientCertificate);
```
Comment (nibanks) [inline src/platform/selfsign_openssl.c:305]: Return not supported for client cert?

File: src/platform/tls_schannel.c [hunk 3]
```diff
@@ -1177,12 +1171,10 @@ CxPlatTlsSecConfigCreate(
...
 
 #ifdef _KERNEL_MODE
-    if (IsClient) {
-        //
-        // Nothing supported for client right now.
-        //
-
-    } else if (CredConfig->Type == QUIC_CREDENTIAL_TYPE_CERTIFICATE_HASH) {
+    //
+    // Assumption: client works the same as server here
```
Comment (nibanks) [inline src/platform/tls_schannel.c:1175]: Assumption? If this is a to-do, can you rephrase this in terms of a `TODO:` comment?

File: src/platform/tls_schannel.c [hunk 4]
```diff
@@ -1793,6 +1785,9 @@ CxPlatTlsWriteDataToSchannel(
         ISC_REQ_CONFIDENTIALITY |
         ISC_RET_EXTENDED_ERROR |
         ISC_REQ_STREAM;
+    if (TlsContext->IsServer) {
+        ContextReq |= ASC_REQ_MUTUAL_AUTH;
+    }
```
Comment (nibanks) [inline src/platform/tls_schannel.c:1790]: Don't we need some kind of config option to indicate that we require a client cert?

File: src/platform/tls_schannel.c [hunk 5]
```diff
@@ -1987,9 +1982,29 @@ CxPlatTlsWriteDataToSchannel(
...
+                        SECPKG_ATTR_REMOTE_CERT_CONTEXT,
+                        (PVOID)&PeerCert);
+#endif
+                if (SecStatus != SEC_E_OK) {
+                    QuicTraceEvent(
+                        TlsErrorStatus,
+                        "[ tls][%p] ERROR, %u, %s.",
+                        TlsContext->Connection,
+                        SecStatus,
+                        "query peer cert");
```
Comment (nibanks) [inline src/platform/tls_schannel.c:2006]:
```suggestion
                        "Query peer cert");
```

File: src/platform/unittest/TlsTest.cpp [hunk 6]
```diff
@@ -29,7 +29,9 @@ struct TlsTest : public ::testing::TestWithParam<bool>
     CXPLAT_SEC_CONFIG* ClientSecConfigCustomCertValidation {nullptr};
     CXPLAT_SEC_CONFIG* ClientSecConfigExtraCertValidation {nullptr};
     CXPLAT_SEC_CONFIG* ClientSecConfigNoCertValidation {nullptr};
+    CXPLAT_SEC_CONFIG* ClientSecConfigClientCert {nullptr};
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:32]: Rename to indicate there is no cert validation:
```suggestion
    CXPLAT_SEC_CONFIG* ClientSecConfigClientCertNoCertValidation {nullptr};
```

File: src/platform/unittest/TlsTest.cpp [hunk 7]
```diff
@@ -55,14 +58,23 @@ struct TlsTest : public ::testing::TestWithParam<bool>
 
     static void SetUpTestSuite()
     {
-        SelfSignedCertParams = CxPlatPlatGetSelfSignedCert(CXPLAT_SELF_SIGN_CERT_USER);
+        SelfSignedCertParams = (QUIC_CREDENTIAL_CONFIG*)CxPlatPlatGetSelfSignedCert(CXPLAT_SELF_SIGN_CERT_USER, FALSE);
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:61]: This cast wasn't needed before. Why now?

File: src/platform/unittest/TlsTest.cpp [hunk 8]
```diff
@@ -55,14 +58,23 @@ struct TlsTest : public ::testing::TestWithParam<bool>
...
+        ASSERT_NE(nullptr, ClientCertParams);
+        ClientCertParams->Flags |= QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION;
+#endif
     }
 
     static void TearDownTestSuite()
     {
         CxPlatPlatFreeSelfSignedCert(SelfSignedCertParams);
         SelfSignedCertParams = nullptr;
+#ifndef QUIC_DISABLE_CLIENT_CERT_TESTS
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:74]: Don't actually need these `ifndef` checks here.

File: src/platform/unittest/TlsTest.cpp [hunk 9]
```diff
@@ -75,6 +87,15 @@ struct TlsTest : public ::testing::TestWithParam<bool>
                 OnSecConfigCreateComplete));
         ASSERT_NE(nullptr, ServerSecConfig);
 
+    SelfSignedCertParams->Flags |= QUIC_CREDENTIAL_FLAG_REQUIRE_CLIENT_AUTHENTICATION;
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:90]:
```suggestion
        SelfSignedCertParams->Flags |= QUIC_CREDENTIAL_FLAG_REQUIRE_CLIENT_AUTHENTICATION;
```

File: src/platform/unittest/TlsTest.cpp [hunk 10]
```diff
@@ -721,7 +760,9 @@ const CXPLAT_TLS_CALLBACKS TlsTest::TlsContext::TlsClientCallbacks = {
     TlsTest::TlsContext::OnPeerCertReceived
 };
 
-const QUIC_CREDENTIAL_CONFIG* TlsTest::SelfSignedCertParams = nullptr;
+QUIC_CREDENTIAL_CONFIG* TlsTest::SelfSignedCertParams = nullptr;
+
+QUIC_CREDENTIAL_CONFIG* TlsTest::ClientCertParams = nullptr;
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:765]:
```suggestion
QUIC_CREDENTIAL_CONFIG* TlsTest::SelfSignedCertParams = nullptr;
QUIC_CREDENTIAL_CONFIG* TlsTest::ClientCertParams = nullptr;
```

File: src/platform/unittest/TlsTest.cpp [hunk 11]
```diff
@@ -128,6 +149,16 @@ struct TlsTest : public ::testing::TestWithParam<bool>
                 &ClientSecConfigNoCertValidation,
                 OnSecConfigCreateComplete));
         ASSERT_NE(nullptr, ClientSecConfigNoCertValidation);
+
+#ifndef QUIC_DISABLE_CLIENT_CERT_TESTS
+        VERIFY_QUIC_SUCCESS(
+            CxPlatTlsSecConfigCreate(
+                ClientCertParams,
+                &TlsContext::TlsClientCallbacks,
+                &ClientSecConfigClientCertNoCertValidation ,
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:158]:
```suggestion
                &ClientSecConfigClientCertNoCertValidation,
```

File: src/platform/unittest/TlsTest.cpp (hunk 7 above)
Comment (anrossi) [reply src/platform/unittest/TlsTest.cpp:61]: Because I'm removing the `const` keyword so I can update the flags field.

Comment (nibanks) [reply src/platform/unittest/TlsTest.cpp:61]: Ok.

File: src/platform/unittest/TlsTest.cpp (hunk 8 above)
Comment (anrossi) [reply src/platform/unittest/TlsTest.cpp:74]: `CxPlatPlatFreeSelfSignedCert` dereferences the parameter to get the `PCCERT_CONTEXT`, which leads to a null deref crash. These are required to avoid that crash.

File: src/platform/unittest/TlsTest.cpp [hunk 12]
```diff
@@ -24,12 +24,15 @@ struct TlsTest : public ::testing::TestWithParam<bool>
...
 protected:
     CXPLAT_SEC_CONFIG* ServerSecConfig {nullptr};
+    CXPLAT_SEC_CONFIG* ServerSecConfigClientAuth {nullptr};
     CXPLAT_SEC_CONFIG* ClientSecConfig {nullptr};
     CXPLAT_SEC_CONFIG* ClientSecConfigDeferredCertValidation {nullptr};
     CXPLAT_SEC_CONFIG* ClientSecConfigCustomCertValidation {nullptr};
     CXPLAT_SEC_CONFIG* ClientSecConfigExtraCertValidation {nullptr};
     CXPLAT_SEC_CONFIG* ClientSecConfigNoCertValidation {nullptr};
-    static const QUIC_CREDENTIAL_CONFIG* SelfSignedCertParams;
+    CXPLAT_SEC_CONFIG* ClientSecConfigClientCertNoCertValidation  {nullptr};
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:33]:
```suggestion
    CXPLAT_SEC_CONFIG* ClientSecConfigClientCertNoCertValidation {nullptr};
```

**Review by nibanks — APPROVED** [review]
